src/queue.ts

```diff
@@ -21,18 +34,29 @@ const startQueue = <T extends EventBase>(
 ) => {
   const queue: number[] = [];
   const dedupeSet = new Set<number>();
+  const processedSet = new Set<number>();
   const promiseMap = new Map<number, PromiseResolver>();
   let state: 'READY' | 'PROCESSING' = 'READY';
 
   const processEvent = (event: EventBase) => {
-    reduxStore.dispatch(event);
     if (event.id === undefined) {
       throw new Error(`Malformed event is missing id: ${event}`);
     }
+    if (processedSet.has(event.id)) {
+      return;
+    }
+    reduxStore.dispatch(event);
     const resolver = promiseMap.get(event.id);
+    processedSet.add(event.id);
     if (resolver) {
       resolver(reduxStore.getState());
-      promiseMap.delete(event.id);
+    }
+
+    // Cleanup
+    promiseMap.delete(event.id);
+    if (event.id % 100_000 === 0) {
+      reduceSetSize(processedSet);
```

Comment:
So every 100,000 events, we remove the earliest 10,000 to save memory?

Reply:
We remove the earliest 90,000 -- yes. It would take 1,000,000 events to consume 100mb of ram, but I also didn't want to wait that long to vacuum.

I had actually meant to put this in a separate PR, but accidentally committed it